### PR TITLE
Saas 13.4 fix livechat seb

### PR DESCRIPTION
**Pull request by seb-odoo**
task-2323670

### CODE REVIEW

File: addons/im_livechat/models/mail_channel.py [hunk 1]
```diff
@@ -89,7 +89,7 @@ def channel_info(self, extra_info=False):
                 channel_infos_dict[channel.id]['livechat_visitor'] = channel._channel_get_livechat_visitor_info()
                 last_msg = self.env['mail.message'].search([("channel_ids", "in", [channel.id])], limit=1)
                 if last_msg:
-                    channel_infos_dict[channel.id]['last_message_date'] = last_msg.date
+                    channel_infos_dict[channel.id]['last_message_id'] = last_msg.id
```
Comment (seb-odoo) [inline addons/im_livechat/models/mail_channel.py:92]: need python test

File: addons/im_livechat/models/mail_channel.py [hunk 2]
```diff
@@ -100,21 +100,28 @@ def channel_fetch_slot(self):
...
 
     def _channel_get_livechat_visitor_info(self):
-        partners = self.channel_partner_ids - self.livechat_operator_id
-        if partners:
+        self.ensure_one()
+        # remove active test to ensure public partner is taken into account
+        channel_partner_ids = self.with_context(active_test=False).channel_partner_ids
+        partners = channel_partner_ids - self.livechat_operator_id
+        if not partners:
+            # operator probably testing the livechat with his own user
```
Comment (seb-odoo) [inline addons/im_livechat/models/mail_channel.py:108]: need python test

File: addons/im_livechat/models/mail_channel.py [hunk 3]
```diff
@@ -100,21 +100,28 @@ def channel_fetch_slot(self):
...
+            partners = channel_partner_ids
+        if partners and partners[0] != self.env.ref('base.public_partner'):
+            # legit non-public partner
             return {
-                'country': partners[0].country_id.name_get() if partners[0].country_id else False,
+                'country': partners[0].country_id.name_get()[0] if partners[0].country_id else False,
```
Comment (seb-odoo) [inline addons/im_livechat/models/mail_channel.py:113]: need python test

File: addons/im_livechat/models/mail_channel.py [hunk 4]
```diff
@@ -100,21 +100,28 @@ def channel_fetch_slot(self):
...
                 'id': partners[0].id,
                 'name': partners[0].name,
             }
         if self.anonymous_name:
             return {
-                'country': False,
+                'country': self.country_id.name_get()[0] if self.country_id else False,
```
Comment (seb-odoo) [inline addons/im_livechat/models/mail_channel.py:119]: need python test

File: addons/im_livechat/models/mail_channel.py [hunk 5]
```diff
@@ -100,21 +100,28 @@ def channel_fetch_slot(self):
...
+            partners = channel_partner_ids
+        if partners and partners[0] != self.env.ref('base.public_partner'):
+            # legit non-public partner
             return {
-                'country': partners[0].country_id.name_get() if partners[0].country_id else False,
+                'country': partners[0].country_id.name_get()[0] if partners[0].country_id else False,
                 'id': partners[0].id,
                 'name': partners[0].name,
             }
         if self.anonymous_name:
```
Comment (seb-odoo) [inline addons/im_livechat/models/mail_channel.py:117]: move this as a condition on name below, don't need 2 returns

File: addons/im_livechat/static/src/components/discuss/discuss_tests.js [hunk 6]
```diff
@@ -69,6 +74,170 @@ QUnit.test('livechat in the sidebar: basic rendering', async function (assert) {
...
+    );
+    assert.containsOnce(
+        document.body,
+        '.o_DiscussSidebar_groupLivechat .o_DiscussSidebar_item',
+        "should have a livechat in the sidebar after receiving first message"
+    );
+    assert.strictEqual(
+        document.querySelector('.o_DiscussSidebar_groupLivechat .o_DiscussSidebar_item').textContent,
+        "Visitor (Belgium)",
+        "should have visitor name as livechat name"
```
Comment (seb-odoo) [inline addons/im_livechat/static/src/components/discuss/discuss_tests.js:161]: and country (anonymous_name)

File: addons/im_livechat/static/src/components/discuss_sidebar/discuss_sidebar.js [hunk 7]
```diff
@@ -19,11 +19,21 @@ patch(components.DiscussSidebar, 'im_livechat/static/src/components/discuss_side
...
-            thread.isPinned &&
-            thread.model === 'mail.channel'
-        );
+        const allOrderedAndPinnedLivechats = this.env.models['mail.thread']
+            .all(thread =>
+                thread.channel_type === 'livechat' &&
+                thread.isPinned &&
+                thread.model === 'mail.channel'
+            ).sort((c1, c2) => {
+                // sort by: lass message id (desc), id (desc)
```
Comment (seb-odoo) [inline addons/im_livechat/static/src/components/discuss_sidebar/discuss_sidebar.js:28]: last

File: addons/im_livechat/static/src/components/discuss_sidebar/discuss_sidebar.js [hunk 8]
```diff
@@ -19,11 +19,21 @@ patch(components.DiscussSidebar, 'im_livechat/static/src/components/discuss_side
...
+                if (c1.lastMessage && c2.lastMessage) {
+                    return c2.lastMessage.id - c1.lastMessage.id;
+                }
+                // a channel without a last message is assumed to be a new
+                // channel just created with the intent of posting a new
+                // message on it, in which case it should be moved up.
+                return c2.id - c1.id;
```
Comment (seb-odoo) [inline addons/im_livechat/static/src/components/discuss_sidebar/discuss_sidebar.js:35]: test this case

File: addons/im_livechat/static/src/models/thread/thread.js [hunk 9]
```diff
@@ -50,6 +61,9 @@ registerInstancePatchModel('mail.thread', 'im_livechat/static/src/models/thread/
      */
     _computeDisplayName() {
         if (this.channel_type === 'livechat' && this.correspondent) {
+            if (this.correspondent.country) {
+                return `${this.correspondent.nameOrDisplayName} (${this.correspondent.country.name})`;
```
Comment (seb-odoo) [inline addons/im_livechat/static/src/models/thread/thread.js:65]: add test for country with non-public user

File: addons/mail/static/tests/chatter_tests.js [hunk 10]
```diff
@@ -133,12 +133,14 @@ QUnit.test('list activity widget with no activity', async function (assert) {
...
-    const lastUserIndex = this.data['res.users'].records.length - 1;
-    this.data['res.users'].records[lastUserIndex].activity_ids = [1, 4];
-    this.data['res.users'].records[lastUserIndex].activity_state = 'today';
-    this.data['res.users'].records[lastUserIndex].activity_summary = 'Call with Al';
-    this.data['res.users'].records[lastUserIndex].activity_type_id = 3;
-    this.data['res.users'].records[lastUserIndex].activity_type_icon = 'fa-phone';
+    const currentUser = this.data['res.users'].records.find(user =>
+        user.id === this.data.currentUserId
+    );
+    currentUser.activity_ids = [1, 4];
```
Comment (seb-odoo) [inline addons/mail/static/tests/chatter_tests.js:139]: object assign (all this file)

**Review by seb-odoo — COMMENTED** [review]
quick check

File: addons/mail/static/tests/helpers/mock_server.js [hunk 11]
```diff
@@ -1032,16 +1224,71 @@ MockServer.include({
...
+                    'res.partner',
+                    [['id', '=', author_id]],
+                    { active_test: false },
+                )[0];
+                email_from = `${author.display_name} <${author.email}>`;
+            }
+        }
+
+        if (!email_from) {
+            throw Error("Unable to log message");
```
Comment (alexkuhn) [inline addons/mail/static/tests/helpers/mock_server.js:1275]: Would clarify what is wrong (something to do with lacking recipient info)
Also technically not just a "log" but ok in that it broadens to the general concept of "history logs" (and it matches wording of error in server code).

File: addons/mail/static/tests/helpers/mock_server.js [hunk 12]
```diff
@@ -1032,16 +1224,71 @@ MockServer.include({
...
         });
     },
+    /**
+     * Simulates `_message_compute_author` on `mail.thread`.
+     *
+     * @private
+     * @param {string} model
+     * @param {integer[]} ids
+     * @param {Object} [context={}]
+     * @return {integer}
```
Comment (alexkuhn) [inline addons/mail/static/tests/helpers/mock_server.js:1234]: Wrong, should be array

File: addons/im_livechat/static/src/components/discuss/discuss_tests.js [hunk 13]
```diff
@@ -69,6 +74,278 @@ QUnit.test('livechat in the sidebar: basic rendering', async function (assert) {
...
+        "should have a livechat in sidebar"
+    );
+    assert.strictEqual(
+        livechat.textContent,
+        "Jean (Belgium)",
+        "should have user name and country as livechat name"
+    );
+});
+
+QUnit.test('add livechat in the sidebar on visitor sending first message', async function (assert) {
```
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/im_livechat/static/src/components/discuss/discuss_tests.js:114]: This is the exact real flow though, it's not possible for a visitor to send a "first message" without "opening his chat" and "sending typing notification" before.

I don't mind "also" adding smaller tests, but there must be one complete, so I'm afraid it's just too much duplicate then.

File: addons/im_livechat/static/src/components/discuss/discuss_tests.js [hunk 14]
```diff
@@ -69,6 +74,278 @@ QUnit.test('livechat in the sidebar: basic rendering', async function (assert) {
...
+    );
+    assert.strictEqual(
+        livechats[1].dataset.threadLocalId,
+        livechat12.localId,
+        "second livechat should now be the one on which the message was not posted"
+    );
+});
+
+
+QUnit.test('livechats in the sidebar sorted with channels without messages at the top', async function (assert) {
```
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/im_livechat/static/src/components/discuss/discuss_tests.js:279]: ok

File: addons/im_livechat/static/src/models/thread/thread.js [hunk 15]
```diff
@@ -17,12 +17,23 @@ registerClassPatchModel('mail.thread', 'im_livechat/static/src/models/thread/thr
...
             if (!data2.members) {
                 data2.members = [];
             }
             if (!data.livechat_visitor.id) {
-                // Create partner derived from public partner.
+                // Create partner derived from public partner and replace the
+                // public partner. The issue with public partner is that it
+                // would be the same record for every livechat, whereas every
+                // correspondent should actually have its own visitor name,
+                // typing status, etc.
```
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/im_livechat/static/src/models/thread/thread.js:35]: Ok, specified that `livechat_visitor` without `id` is the anonymous visitor.

File: addons/im_livechat/static/src/models/thread/thread.js [hunk 16]
```diff
@@ -17,12 +17,23 @@ registerClassPatchModel('mail.thread', 'im_livechat/static/src/models/thread/thr
...
+                data2.members.push(['unlink', this.env.messaging.publicPartner]);
```
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/im_livechat/static/src/models/thread/thread.js:36]: ok, "briefly" not possible when the comment is already 5 lines long, but I explained everything

File: addons/mail/static/src/models/thread/thread.js [hunk 17]
```diff
@@ -792,9 +792,14 @@ function factory(dependencies) {
             const correspondents = this.members.filter(partner =>
                 partner !== this.env.messaging.currentPartner
             );
+            // 2 members chat
             if (correspondents.length === 1) {
                 return [['link', correspondents[0]]];
             }
+            // chat with oneself
```
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/mail/static/src/models/thread/thread.js:799]: ok

File: addons/mail/static/tests/helpers/mock_server.js [hunk 18]
```diff
@@ -445,11 +518,47 @@ MockServer.include({
...
+     * @return {Object}
+     */
+    _mockMailChannel_channelChannelNotifications(ids, partner_ids) {
+        const notifications = [];
+        for (const partner_id of partner_ids) {
+            const user = this._getRecords('res.users', [['partner_id', 'in', partner_id]])[0];
+            if (!user) {
+                continue;
+            }
+            // TODO should call channel_info with the proper user context
```
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/mail/static/tests/helpers/mock_server.js:548]: Good catch, I'll just remove TODO and add a note instead, comment doesn't apply if we don't have mocked channel.partner

File: addons/mail/static/tests/helpers/mock_server.js [hunk 19]
```diff
@@ -1032,16 +1224,71 @@ MockServer.include({
...
+     * @private
+     * @param {string} model
+     * @param {integer[]} ids
+     * @param {Object} [context={}]
+     * @return {integer}
+     */
+    _MockMailThread_MessageComputeAuthor(model, ids, author_id, email_from, context = {}) {
+        if (author_id === undefined) {
+            // For simplicity, partner is not guessed from email_from here, but
+            // that is would be the first step on the server.
```
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/mail/static/tests/helpers/mock_server.js:1239]: indeed, "that would be"

File: addons/mail/static/tests/helpers/mock_server.js (hunk 11 above)
Comment (seb-odoo) [reply addons/mail/static/tests/helpers/mock_server.js:1275]: It's missing author info actually not recipient info, I'll add that to the error message.

File: addons/mail/static/tests/helpers/mock_server.js (hunk 12 above)
Comment (seb-odoo) [reply addons/mail/static/tests/helpers/mock_server.js:1234]: good catch, should be `returns` as well, not sure from which I copied it but I got it wrong everywhere, I'll change them all.

**Comment by seb-odoo** [comment]
Changes done, also updated all commits to include PR and task.
Now need to rebase and resolve conflict.

File: addons/im_livechat/static/src/components/discuss/discuss_tests.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/im_livechat/static/src/components/discuss/discuss_tests.js:114]: Please check again the PR. All comments done except this one. Waiting response to do anything.

File: addons/im_livechat/static/src/models/thread/thread.js [hunk 20]
```diff
@@ -17,12 +17,35 @@ registerClassPatchModel('mail.thread', 'im_livechat/static/src/models/thread/thr
...
+                 *
+                 * But the issue with public partner is that it is the same
+                 * record for every livechat, whereas every correspondent should
+                 * actually have its own visitor name, typing status, etc.
+                 *
+                 * Due to JS being temporary by nature there is no such notion
+                 * of polluting the database, it is therefore acceptable and
+                 * easier to handle one temporary partner per channel.
+                 */
+                data2.members.push(['unlink', this.env.messaging.publicPartner]);
```
Comment (alexkuhn) [inline addons/im_livechat/static/src/models/thread/thread.js:48]: 👍

File: addons/im_livechat/static/src/components/discuss/discuss_tests.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (alexkuhn) [reply addons/im_livechat/static/src/components/discuss/discuss_tests.js:114]: Test is a bit too long IMO, and it makes use of typing notification feature when it shouldn't be... Else the more features we have, the more verbose the existing tests they become.

I would prefer one test just for `get_session` not pinning livechat, one test for `typing notif` not pinning livechat initially, and one test for `chat_post` that actually pin the livechat.

The last test doesn't need as many assertions. And it's nicer to keep qunit tests as small as possible.
True it's annoying not being able to define dependencies among tests with QUnit, but it's worse turning tests into long flows.

Ideally, each test should basically be:

- setup phase
- (optional assertion(s))
- action(s)
- assertion(s)

Having tests with such shape is a sign that it does too much and should be split in smaller tests:
`action(s) 1 => assertion(s) 1 => action(s) 2 => assertion(s) 2 => ...`

[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/im_livechat/static/src/components/discuss/discuss_tests.js]: done

File: addons/im_livechat/static/src/models/thread/thread.js [hunk 21]
```diff
@@ -17,12 +17,35 @@ registerClassPatchModel('mail.thread', 'im_livechat/static/src/models/thread/thr
      */
     convertData(data) {
         const data2 = this._super(data);
+        if ('last_message_id' in data) {
+            if (!data2.messagesAsServerChannel) {
+                data2.messagesAsServerChannel = [];
+            }
+            data2.messagesAsServerChannel.push(['insert', { id: data.last_message_id }]);
+        }
```
Comment (alexkuhn) [inline addons/im_livechat/static/src/models/thread/thread.js:25]: This is a bit annoying that we have to update ultra-specific fields in order to manipulate other fields that feel much more natural...
Wouldn't change anything in this PR, just a reminder that even if computed fields is an improvement over none, it's still cumbersome to manage `x2many` relations containing groups...

**Review by alexkuhn — APPROVED** [review]
@robodoo r+ rebase-ff

**Pull request closed without merging**